### PR TITLE
benchmark: add lightning import for TPC-C

### PR DESCRIPTION
**Pull request by jackysp**
Signed-off-by: Shuaipeng Yu <jackysp@gmail.com>

### What is changed, added or deleted? 
Add lightning import for TPC-C.

### Which version does your change affect? 
dev, 3.0, 3.1

### CODE REVIEW

**Comment by jackysp** [comment]
I'll update 3.0 and 3.1 when it is ready to merge.

File: dev/benchmark/how-to-run-tpcc.md [hunk 1]
```diff
@@ -149,6 +153,88 @@ cd run && \
...
+[importer_server]
+IS1 ansible_host=172.16.5.34 deploy_dir=/data2/is1 tikv_importer_port=13323 import_dir=/data2/import
+
+[lightning_server]
+LS1 ansible_host=172.16.5.34 deploy_dir=/data2/ls1 tidb_lightning_pprof_port=23323 data_source_dir=/home/user/csv
+```
+
+##### 修改 conf/tidb-lightning.yml
+
+```text
```
Comment (kennytm) [inline dev/benchmark/how-to-run-tpcc.md]:
````suggestion
```yaml
````

File: dev/benchmark/how-to-run-tpcc.md [hunk 2]
```diff
@@ -149,6 +153,88 @@ cd run && \
...
+
+#### 通过 Lightning 导入
+
+通过 Lightning 导入数据请参考 [Lightning 部署执行](dev/reference/tools/tidb-lightning/deployment.md)章节。这里我们介绍下通过 tidb-ansible 部署 Lightning 导入数据的方法。
+
+##### 修改 inventory.ini
+
+这里最好手动指定清楚部署的 IP、端口、目录，避免各种冲突问题带来的异常，其中 import_dir 的磁盘空间参考 [Lightning 部署执行](dev/reference/tools/tidb-lightning/deployment.md)，data_source_dir 就是存储上一节 csv 数据的目录。
+
+```text
```
Comment (kennytm) [inline dev/benchmark/how-to-run-tpcc.md]:
````suggestion
```ini
````

File: dev/benchmark/how-to-run-tpcc.md [hunk 3]
```diff
@@ -149,6 +153,88 @@ cd run && \
...
+./runLoader.sh props.mysql
+```
+
+#### 通过 Lightning 导入
+
+通过 Lightning 导入数据请参考 [Lightning 部署执行](dev/reference/tools/tidb-lightning/deployment.md)章节。这里我们介绍下通过 tidb-ansible 部署 Lightning 导入数据的方法。
+
+##### 修改 inventory.ini
+
+这里最好手动指定清楚部署的 IP、端口、目录，避免各种冲突问题带来的异常，其中 import_dir 的磁盘空间参考 [Lightning 部署执行](dev/reference/tools/tidb-lightning/deployment.md)，data_source_dir 就是存储上一节 csv 数据的目录。
```
Comment (kennytm) [inline dev/benchmark/how-to-run-tpcc.md]:
```suggestion
这里最好手动指定清楚部署的 IP、端口、目录，避免各种冲突问题带来的异常，其中 import_dir 的磁盘空间参考 [Lightning 部署执行](/dev/reference/tools/tidb-lightning/deployment.md)，data_source_dir 就是存储上一节 csv 数据的目录。
```

File: dev/benchmark/how-to-run-tpcc.md [hunk 4]
```diff
@@ -149,6 +153,88 @@ cd run && \
...
+
+{{< copyable "shell-regular" >}}
+
+```shell
+./runLoader.sh props.mysql
+```
+
+#### 通过 Lightning 导入
+
+通过 Lightning 导入数据请参考 [Lightning 部署执行](dev/reference/tools/tidb-lightning/deployment.md)章节。这里我们介绍下通过 tidb-ansible 部署 Lightning 导入数据的方法。
```
Comment (kennytm) [inline dev/benchmark/how-to-run-tpcc.md]:
```suggestion
通过 Lightning 导入数据请参考 [Lightning 部署执行](/dev/reference/tools/tidb-lightning/deployment.md)章节。这里我们介绍下通过 tidb-ansible 部署 Lightning 导入数据的方法。
```

File: dev/benchmark/how-to-run-tpcc.md [hunk 5]
```diff
@@ -149,6 +153,88 @@ cd run && \
...
+```
+
+##### 启动
+
+* 登录到部署 Lightning 和 Importer 的服务器
+* 进入部署目录
+* 在 Importer 目录下创建一个 log 文件夹
+* 在 Importer 目录下执行 `scripts/start_importer.sh`，启动 Importer
+* 在 Lightning 目录下创建一个 log 文件夹
+* 在 Lightning 目录下执行 `scripts/start_lightning.sh`，开始导入数据
```
Comment (kennytm) [inline dev/benchmark/how-to-run-tpcc.md]: log 文件夾應該會隨 shell script 自動建立的 🤔

Comment (jackysp) [reply dev/benchmark/how-to-run-tpcc.md]: It doesn't.

File: dev/benchmark/how-to-run-tpcc.md [hunk 6]
```diff
@@ -149,6 +153,88 @@ cd run && \
...
+
+[lightning_server]
+LS1 ansible_host=172.16.5.34 deploy_dir=/data2/ls1 tidb_lightning_pprof_port=23323 data_source_dir=/home/user/csv
+```
+
+##### 修改 conf/tidb-lightning.yml
+
+```yaml
+tikv_importer:
+    addr: "172.16.5.34:13323" # 这里指向 importer 的地址，需要将上面 inventory.ini 里指定的再写一次
```
Comment (jackysp) [inline dev/benchmark/how-to-run-tpcc.md]: Could we improve this? I mean it doesn't need to set IP:port once again. @kennytm @liubo0127

File: dev/benchmark/how-to-run-tpcc.md (hunk 5 above)
Comment (jackysp) [reply dev/benchmark/how-to-run-tpcc.md]: \cc @liubo0127

File: dev/benchmark/how-to-run-tpcc.md (hunk 6 above)
Comment (kennytm) [reply dev/benchmark/how-to-run-tpcc.md]: It should be fixed in tikv/importer#20.

File: dev/benchmark/how-to-run-tpcc.md [hunk 7]
```diff
@@ -149,6 +153,88 @@ cd run && \
 
 根据机器配置这个过程可能会持续几个小时。
 
+### 通过 Lightning 导入
```
Comment (yikeke) [inline dev/benchmark/how-to-run-tpcc.md]:
```suggestion
### 通过 TiDB Lightning 导入
```

File: dev/benchmark/how-to-run-tpcc.md [hunk 8]
```diff
@@ -149,6 +153,88 @@ cd run && \
...
+
+由于导入数据量随着 warehouse 的增加而增加，当需要导入 1000 warehouse 以上数据时，可以使用先用 BenchmarkSQL 生成 csv 文件，再将文件通过 Ligtning 导入的方式来快速导入。生成的 csv 文件也可以多次复用，节省每次生成所需要的时间。
```
Comment (yikeke) [inline dev/benchmark/how-to-run-tpcc.md]:
```suggestion
由于导入数据量随着 warehouse 的增加而增加，当需要导入 1000 warehouse 以上数据时，可以先用 BenchmarkSQL 生成 csv 文件，再将文件通过 TiDB Lightning（以下简称 Lightning）导入的方式来快速导入。生成的 csv 文件也可以多次复用，节省每次生成所需要的时间。
```

File: dev/benchmark/how-to-run-tpcc.md [hunk 9]
```diff
@@ -149,6 +153,88 @@ cd run && \
...
+
+#### 修改 BenchmarkSQL 的配置文件
+
+1 warehouse 的 csv 文件需要 77 MB 磁盘空间，在生成之前要根据需要分配足够的磁盘空间来保存 csv 文件。修改 `benchmarksql/run/props.mysql`，增加一行：
```
Comment (yikeke) [inline dev/benchmark/how-to-run-tpcc.md]:
```suggestion
1 warehouse 的 csv 文件需要 77 MB 磁盘空间，在生成之前要根据需要分配足够的磁盘空间来保存 csv 文件。可以在 `benchmarksql/run/props.mysql` 文件中增加一行：
```

File: dev/benchmark/how-to-run-tpcc.md [hunk 10]
```diff
@@ -149,6 +153,88 @@ cd run && \
...
+
+```text
+fileLocation=/home/user/csv/  # 存储 csv 文件的目录绝对路径，需保证有足够的空间
+```
+
+因为最终要使用 Lightning 导入数据，因此，csv 文件名最好符合 Lightning 要求，即 {database}.{table}.csv 的命名法。所以，这里可以使用略 trick 的方法将以上配置改为：
```
Comment (yikeke) [inline dev/benchmark/how-to-run-tpcc.md]:
```suggestion
因为最终要使用 Lightning 导入数据，所以 csv 文件名最好符合 Lightning 要求，即 `{database}.{table}.csv` 的命名法。这里可以将以上配置改为：
```
“略trick” is too casual and its style does not fit in our document. Is it okay to delete it?

File: dev/benchmark/how-to-run-tpcc.md [hunk 11]
```diff
@@ -149,6 +153,88 @@ cd run && \
...
+
+```text
+fileLocation=/home/user/csv/tpcc.  # 存储 csv 文件的目录绝对路径 + 文件名前缀（database）
+```
+
+这样生成的 csv 文件将会是 tpcc.bmsql_warehouse.csv 的符合 Lightning 样式。
```
Comment (yikeke) [inline dev/benchmark/how-to-run-tpcc.md]:
```suggestion
这样生成的 csv 文件名将会是类似 `tpcc.bmsql_warehouse.csv` 的样式，符合 Lightning 的要求。
```

File: dev/benchmark/how-to-run-tpcc.md [hunk 12]
```diff
@@ -121,6 +121,8 @@ loadWorkers=32  # 导入数据的并发数
 
 ## 导入数据
 
+**导入数据通常是整个 TPC-C 测试中最耗时，也是最容易出问题的阶段**
```
Comment (yikeke) [inline dev/benchmark/how-to-run-tpcc.md]:
```suggestion
**导入数据通常是整个 TPC-C 测试中最耗时，也是最容易出问题的阶段。**
```

File: dev/benchmark/how-to-run-tpcc.md (hunk 5 above)
Comment (liubo0127) [reply dev/benchmark/how-to-run-tpcc.md]: All directorys should be created automatically. I'll check it.

File: dev/benchmark/how-to-run-tpcc.md (hunk 6 above)
Comment (liubo0127) [reply dev/benchmark/how-to-run-tpcc.md]: It should be configured automatically.

Comment (liubo0127) [reply dev/benchmark/how-to-run-tpcc.md]: I have tested, and it can be configured automatically.

File: dev/benchmark/how-to-run-tpcc.md (hunk 5 above)
Comment (liubo0127) [reply dev/benchmark/how-to-run-tpcc.md]: Fixed in https://github.com/pingcap/tidb-ansible/pull/1103

Comment (jackysp) [reply dev/benchmark/how-to-run-tpcc.md]: great

File: dev/benchmark/how-to-run-tpcc.md (hunk 6 above)
Comment (jackysp) [reply dev/benchmark/how-to-run-tpcc.md]: 企业微信截图_20200110102808
企业微信截图_20200110102653

Comment (jackysp) [reply dev/benchmark/how-to-run-tpcc.md]: port does not set automatically.

Comment (liubo0127) [reply dev/benchmark/how-to-run-tpcc.md]: Ok, it's actually set automatically, but it doesn't match the custom port.

Comment (liubo0127) [reply dev/benchmark/how-to-run-tpcc.md]: Fixed.

Comment (jackysp) [reply dev/benchmark/how-to-run-tpcc.md]: great, thx

**Comment by jackysp** [comment]
All comments addressed, and apply to v3.0 and v3.1, PTAL @yikeke

**Comment by jackysp** [comment]
... the bot blocks CLA.

**Comment by yikeke** [comment]
> ... the bot blocks CLA.

I asked the bot's manager to sign it. Wait a moment.

**Comment by you06** [comment]
Facing a bot-login issue, we're trying to solve this.

File: dev/benchmark/how-to-run-tpcc.md (hunk 1 above)
Comment (july2993) [reply dev/benchmark/how-to-run-tpcc.md]: 不是用的 toml  的吗，都兼容？

File: dev/benchmark/how-to-run-tpcc.md [hunk 13]
```diff
@@ -149,6 +153,84 @@ cd run && \
...
+
+```ini
+[importer_server]
+IS1 ansible_host=172.16.5.34 deploy_dir=/data2/is1 tikv_importer_port=13323 import_dir=/data2/import
+
+[lightning_server]
+LS1 ansible_host=172.16.5.34 deploy_dir=/data2/ls1 tidb_lightning_pprof_port=23323 data_source_dir=/home/user/csv
+```
+
+##### 修改 conf/tidb-lightning.yml
```
Comment (july2993) [inline dev/benchmark/how-to-run-tpcc.md]: not `toml`?  @kennytm

Comment (kennytm) [reply dev/benchmark/how-to-run-tpcc.md]: yes toml 😓

Comment (kennytm) [reply dev/benchmark/how-to-run-tpcc.md]: YAML is correct, Ansible uses YAML as template. 😓 @july2993

Comment (july2993) [reply dev/benchmark/how-to-run-tpcc.md]: I methinks this is the config file `lightning` use directly  😓

**Comment by jackysp** [comment]
The issue found in this PR is fixed by https://github.com/pingcap/tidb-ansible/pull/1105, https://github.com/pingcap/tidb-ansible/pull/1104 and https://github.com/pingcap/tidb-ansible/pull/1103

**Pull request merged**